ElfLinker in LGC

trenouf opened the pull request:
Only the top four commits are part of this PR:

2cf5d75c0 (HEAD) Switch Compiler::linkRelocatableShaderElf to use new LGC ElfLinker API
7ba69668b Implemented LGC ElfLinker API
03bfd629a Define constants for relocs
44a8a807d Added BuilderBase::CreateAddByteOffset

This will fail to build for now because it needs an LLVM patch:
https://reviews.llvm.org/D79671

trenouf commented:
Rebased

trenouf commented:
V2: Fixed cantFail problem on earlier LLVM

trenouf commented:
"Fast Jenkins Build" failed because it is using the wrong LLVM. I pushed the required LLVM fix to GPUOpen branch amd-gfx-gpuopen-dev, but I guess that build is not using it.

Flakebi commented:
LLVM and other parts are cached in this build and only rebuilt once a day, same as the GitHub Matrix builds (the last one ran 20h ago). I started a new base build, which should finish in ~30 minutes.

Flakebi commented:
retest this please

amdvlk-admin commented:
### Test summary for commit 7c4e897c7c917c0da99581a15b5e63183513e4bc
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev ddcaa1009ecde219613042a5a53c6d3752bf4b9d</li>
<li>SPVGEN: dev d4817ab957f79762da83b7fc1d3f2816dbdc13fd</li>
<li>XGL: dev 2ecd2e14e2ae4f2712df4a2220e2dd36265c66d1</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c57487b76215ee74b1038039f47b210a28cad65e</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103198/103199 (100.0%)</li>
<li><details><summary>Failed: 1/103199 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.object_management.multithreaded_shared_resources.graphics_pipeline
Stack trace: result: 0xfffffffffffffff3 at vktApiObjectManagementTests.cpp:1827
```
</p></details></li>


</ul></details>


</ul></details>

trenouf commented:
Spurious failure on one of three CTS runs.

trenouf commented:
Rebased.

trenouf commented on lgc/builder/BuilderBase.cpp:
```diff
@@ -78,3 +78,30 @@ Value *BuilderBase::CreateRelocationConstant(const Twine &symbolName) {
+// @param pointer : Pointer to add to
+// @param byteOffset : Byte offset to add
+// @param instName : Name to give instruction
+Value *BuilderBase::CreateAddByteOffset(Value *pointer, Value *byteOffset, const Twine &instName) {
```
This is only needed when byteOffset might be a reloc, and only then to work around a back-end problem. In general, people can just use GEP.

trenouf commented on lgc/include/lgc/state/AbiUnlinked.h:
```diff
@@ -0,0 +1,91 @@
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  AbiUnlinked.h
```
"Unlinked" is the term I am using for separately compiling a shader or half-pipeline, and then later linking them. There is a comment here to briefly explain the purpose of the file:

 * This file contains declarations for the ABI of an unlinked shader or half-pipeline ELF, generated by LGC
 * compilation and consumed by the LGC linker.

There will be further stuff added to this file, including some PAL metadata keys.

nhaehnle commented on lgc/elfLinker/ElfLinker.cpp:
```diff
@@ -0,0 +1,645 @@
+        // All text and rodata sections get lumped together, even if they have different names.
+        m_outputSections[textSectionIdx].addInputSection(elfInput, section);
+      } else if (elfSection.getType() != ELF::SHT_NOTE) {
+        // Put same-named sections together, and exclude .note.
+        StringRef name = cantFail(section.getName());
+        if (name.startswith(".AMDGPU.")) {
+          for (unsigned idx = 1;; ++idx) {
+            if (idx == m_outputSections.size()) {
+              m_outputSections.push_back(OutputSection(this));
+              m_outputSections[idx].addInputSection(elfInput, section);
+              break;
+            }
+            if (name == m_outputSections[idx].getName()) {
+              m_outputSections[idx].addInputSection(elfInput, section);
+              break;
+            }
+          }
+        }
```
Why are we discarding sections whose names do not start with ".AMDGPU."?

trenouf replied:
It was an attempt to exclude things like the symbol table and reloc sections. But I'll change it to do that in a better way...

trenouf replied:
I thought I resolved this, but I haven't.

nhaehnle replied:
Any resolution on this? The only thing "special" PROGBITS thing I see in a typical ELF is .AMDGPU.disasm. symtab and strtab aren't of type PROGBITS according to llvm-readelf.

trenouf replied:
Yeah, sorry, I have a resolution for this locally (don't check for .AMDGPU at all, and just use PROGBITS), but I haven't got round to pushing it yet.

trenouf replied:
Really done it this time.

nhaehnle commented on lgc/elfLinker/ElfLinker.cpp:
```diff
@@ -0,0 +1,645 @@
+  const size_t paddingUnit = 16;
+  const char *padding = "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0";
+  bool paddingAtEnd = false;
+  if (shdr->sh_flags & ELF::SHF_EXECINSTR) {
+    if (m_linker->getPipelineState()->getTargetInfo().getGfxIpVersion().major < 10)
+      padding = "\0\0\x80\xBF\0\0\x80\xBF\0\0\x80\xBF\0\0\x80\xBF"; // s_nop
+    else {
+      padding = "\0\0\x9F\xBF\0\0\x9F\xBF\0\0\x9F\xBF\0\0\x9F\xBF"; // s_code_end
+      paddingAtEnd = true;
+    }
+  }
```
Presumably this needs adjustment when we start having glue code? There's still reason to align to a higher multiple (e.g. for loop alignment), but obviously it needs an s_nop padding between glue parts on all hardware...

trenouf replied:
Yes, I think you're right. Glue code also needs additional handling: it needs to remove the final s_endpgm in some way (which also requires constant pools to be in a different rodata section, not done yet).

nhaehnle replied:
We should just use the existing "return" facility for that: for Mesa support, LLVM already knows not to emit `s_endpgm` for shaders that return non-void, and it knows to put the return basic block last in ISA order. (The SI_RETURN_TO_EPILOG pseudo-instruction is used in that case.)

trenouf replied:
Yeah, I spotted that whilst working on the fetch shader, and I assumed it was a deliberate policy for Mesa in the LLVM back-end (without actually checking).

But it still seems to add the s_code_end padding for GFX10, which is a shame. Do you think we should fix the back-end to not do that?

nhaehnle replied:
Yes, fixing the backend for that would be good. I think it's already fixed for Mesa, but contingent on the OS in the triple. For the amdpal OS, the question boils down to whether we want to build a relocatable ELF or not... I haven't thought about how to best express that. Of course, we could just heuristically scan if there are any non-void-returning functions in the module and decide based on that...

trenouf replied:
The back-end fix is D80338, in review.

nhaehnle commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -1031,12 +1037,14 @@ void PatchEntryPointMutate::addUserDataArgs(SmallVectorImpl<UserDataArg> &userDa
+      if (m_pipelineState->getShaderOptions(m_shaderStage).updateDescInElf && m_shaderStage == ShaderStageFragment) {
         // Put set number to register first, will update offset after merge ELFs
-        // For partial pipeline compile, only fragment shader needs to adjust offset of root descriptor
-        // If there are more individual shader compile in future, we can add more stages here
+        // For partial pipeline compile, only fragment shader needs to adjust offset of root descriptor.
+        // This is part of the original "partial pipeline compile" scheme, and it uses a magic number for the
```
Is this claim about "only the fragment shader" actually true? If we don't have the pipeline layout for a partial, relocatable compile, then the same would apply to vertex/geometry shaders, no?

trenouf replied:
Maybe, but that is part of Qun and DavidZ's "partial pipeline compile" scheme, which I'm trying not to touch here. In the condition above, I removed the checks for vertex and compute shader to restore that bit of functionality to how it was left by DavidZ. In our new "shader/half-pipeline compile", we don't use this any more.

nhaehnle commented on lgc/state/PalMetadata.cpp:
```diff
@@ -240,11 +295,89 @@ void PalMetadata::setUserDataSpillUsage(unsigned dwordOffset) {
+// Fix up user data registers. Any user data register that has one of the unlinked UserDataMapping values defined
+// in AbiUnlinked.h is fixed up by looking at pipeline state.
+void PalMetadata::fixUpRegisters() {
+  static const unsigned ComputeRegRanges[] = {mmCOMPUTE_USER_DATA_0, 16};
+  static const unsigned Gfx8RegRanges[] = {mmSPI_SHADER_USER_DATA_PS_0, 16, mmSPI_SHADER_USER_DATA_VS_0, 16,
+                                           mmSPI_SHADER_USER_DATA_GS_0, 16, mmSPI_SHADER_USER_DATA_ES_0, 16,
+                                           mmSPI_SHADER_USER_DATA_HS_0, 16, mmSPI_SHADER_USER_DATA_LS_0, 16};
+  static const unsigned Gfx9RegRanges[] = {mmSPI_SHADER_USER_DATA_PS_0, 32, mmSPI_SHADER_USER_DATA_VS_0, 32,
+                                           mmSPI_SHADER_USER_DATA_ES_0, 32, mmSPI_SHADER_USER_DATA_HS_0, 32};
+  static const unsigned Gfx10RegRanges[] = {mmSPI_SHADER_USER_DATA_PS_0, 32, mmSPI_SHADER_USER_DATA_VS_0, 32,
```
Maybe arrays of `std::pair<unsigned, unsigned>` with nested-braces for initialization? A quick check suggests that that also gets emitted as .rodata.

trenouf commented:
V3: Review comments: detection of which sections to include, and
        std::pair initialization on reg ranges to fix up.

amdvlk-admin commented:
### Test summary for commit 7c2718b82ec1ebec208412ef3850951bc91f85dc
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 85170384d0ee5447ac267d478f7a21bc036ea71c</li>
<li>SPVGEN: dev d4817ab957f79762da83b7fc1d3f2816dbdc13fd</li>
<li>XGL: dev e3442a6897c2344bb147e225622a108c3c767fbe</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c57487b76215ee74b1038039f47b210a28cad65e</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 34117/34118 (100.0%)</li>
<li><details><summary>Failed: 1/34118 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.binding_model.shader_access.secondary_cmd_buf.combined_image_sampler_immutable.fragment.multiple_descriptor_sets.multiple_discontiguous_descriptors.2d_base_mip
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>


</ul></details>

trenouf commented:
V4: Removed accidentally left behind empty else

amdvlk-admin commented:
### Test summary for commit 5490669f76ee54111e77aafa52c1cb7732e0412d
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 85170384d0ee5447ac267d478f7a21bc036ea71c</li>
<li>SPVGEN: dev 704f4adddb6867d633431c2c9c2c0bfa6d294ce2</li>
<li>XGL: dev 9c4fe791366e730131a0d99a7519855068f216ef</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c57487b76215ee74b1038039f47b210a28cad65e</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 2083/2084 (100.0%)</li>
<li><details><summary>Failed: 1/2084 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.object_management.max_concurrent.command_buffer_primary
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>


</ul></details>

trenouf commented:
V5: Better detection of which sections to include. Moved PAL metadata
        gathering earlier. Changed to newer MsgPack merge API. A few other
        changes in preparation for fetch shader. Removed unnecessary
        msgpack::Document::getNode calls.

Also added new commit "lgc -l (link) and -glue options".

trenouf commented:
V5: Better detection of which sections to include. Moved PAL metadata
        gathering earlier. Changed to newer MsgPack merge API. A few other
        changes in preparation for fetch shader. Removed unnecessary
        msgpack::Document::getNode calls.

Also added new commit "lgc -l (link) and -glue options".

trenouf commented:
V5: Better detection of which sections to include. Moved PAL metadata
        gathering earlier. Changed to newer MsgPack merge API. A few other
        changes in preparation for fetch shader. Removed unnecessary
        msgpack::Document::getNode calls.

Also added new commit "lgc -l (link) and -glue options".

amdvlk-admin commented:
### Test summary for commit 1479b7006fa78f23f3f0b630c77cff34f9d9cfe9
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 85170384d0ee5447ac267d478f7a21bc036ea71c</li>
<li>SPVGEN: dev 704f4adddb6867d633431c2c9c2c0bfa6d294ce2</li>
<li>XGL: dev 9c4fe791366e730131a0d99a7519855068f216ef</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 3a1103f746f4d782271bd32d67274d6260c5dc7f</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 959/959 (100.0%)</li>
<li>Failed: 0/959 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 35498/35499 (100.0%)</li>
<li><details><summary>Failed: 1/35499 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.binding_model.descriptorset_random.sets32.constant.ubolimithigh.sbolimithigh.imglimithigh.noiub.frag.ialimithigh.0
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>


</ul></details>

trenouf commented:
Ping for review/approval.

trenouf commented:
V6: Do not set pipeline type or VS HW stage in metadata when compiling FS only.

Also fixed lgc output to stdout when input is stdin.

amdvlk-admin commented:
### Test summary for commit fdad26ad7f67f8a1f09c7c3b8080f1115ddca7ac
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 85170384d0ee5447ac267d478f7a21bc036ea71c</li>
<li>SPVGEN: dev 704f4adddb6867d633431c2c9c2c0bfa6d294ce2</li>
<li>XGL: dev 9c4fe791366e730131a0d99a7519855068f216ef</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 3a1103f746f4d782271bd32d67274d6260c5dc7f</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>


</ul></details>

s-perron commented on lgc/state/PalMetadata.cpp:
```diff
@@ -237,19 +295,102 @@ void PalMetadata::setUserDataEntry(ShaderStage stage, unsigned userDataIndex, un
+  SmallVector<const ResourceNode *, 4> descSetNodes;
+  const ResourceNode *pushConstNode;
+  for (const auto &node : m_pipelineState->getUserDataNodes()) {
+    if (node.type == ResourceNodeType::DescriptorTableVaPtr && !node.innerTable.empty()) {
```
This seems to assume that descriptor sets are not in the top level.  However, xgl creates the internal pipelines with descriptor sets in the top level.  See https://github.com/GPUOpen-Drivers/xgl/blob/dev/icd/api/vk_device.cpp#L1931.

Is there any plan to handle this?  This is causes all applications to fail when using relocatable shaders because this compute shader is created for every application.

trenouf replied:
Hmm, good point, and now you mention it, I think I recall that Rex mentioned it too.

Can we avoid using shader compilation on those internal pipelines? Presumably they are amenable to pipeline caching.

Alternatively, and I think this is a question to Qun and Rex, can we stop putting descriptor sets in the top level in internal shaders?

trenouf replied:
Also, we should make ElfLinker::link() spot the situation and return a recoverable error, and enhance llpcCompiler to do something with a recoverable error (chuck away the shader elf(s) and do a pipeline compile).

s-perron replied:
Currently there is no way for the driver to specify which shaders should be compiled using relocatable shaders and which ones should not.  It is a single global option to LLPC.  I've been thinking that we might be able to change that by adding a new option in the pipeline state or something.  We can discuss it later.

We should add something to `canUseRelocatableShader*Elf` to say that this pipeline cannot be compiled using relocatable shaders, and then it will fall back to building the full pipeline.  Anything else we do should be on top of this.

amdrexu replied:
Qun will make comments for this.

linqun replied:
> Hmm, good point, and now you mention it, I think I recall that Rex mentioned it too.
> 
> Can we avoid using shader compilation on those internal pipelines? Presumably they are amenable to pipeline caching.
> 
> Alternatively, and I think this is a question to Qun and Rex, can we stop putting descriptor sets in the top level in internal shaders?

I think it is possible now. until now, we only have one internal compute shader in vk::Device.  please check Device::CreateInternalPipelines (vk_device.cpp) for detail.
but we may encounter more issues for ray-tracing internal pipelines in the future.  please check related internal source code (gpurtInternalShaders.cpp) for detail.

s-perron replied:
I've opened https://github.com/GPUOpen-Drivers/llpc/pull/723 to get this working.

nhaehnle replied:
FWIW, I'd say having a per-pipeline setting to enable/disable this feature is something we probably want anyway in the long run, to be exposed to applications via the Vulkan API. But that's unrelated to this change.

trenouf commented:
V7: Ensure .note section is 4-aligned in file to avoid undefined
        behavior in vkgcPipelineDumper.
        Remove gfx10 s_code_end_padding from linked .text sections.

trenouf commented:
V8: Fixed compile error with std::min.

amdvlk-admin commented:
### Test summary for commit 86d2b6d358317493030ffbc4766c4a45ede97216
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 85170384d0ee5447ac267d478f7a21bc036ea71c</li>
<li>SPVGEN: dev 704f4adddb6867d633431c2c9c2c0bfa6d294ce2</li>
<li>XGL: dev 9c4fe791366e730131a0d99a7519855068f216ef</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 3a1103f746f4d782271bd32d67274d6260c5dc7f</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>


</ul></details>
